### DISCUSSION TITLE
Russia support

### DISCUSSION THREAD

**Discussion by romanovroman74** [Hyundai-Kia-Connect/kia_uvo#474] [Ideas]
**Region and Brand of car**
EU/Russia, Hyundai Tucson 2021

**Version of the integration**
2.0.40
Home Assistant version: 2022.11.5

**Describe the bug**
Unable to login to integration

**Debug logs if an error occurred**

Logger: custom_components.kia_uvo.config_flow
Source: custom_components/kia_uvo/config_flow.py:170 
Integration: Kia Uvo / Hyundai Bluelink ([documentation](https://github.com/fuatakgun/kia_uvo), [issues](https://github.com/fuatakgun/kia_uvo/issues)) 
First occurred: 12:26:47 (7 occurrences) 
Last logged: 12:43:34

kia_uvo Exception in kia_uvo login : can only concatenate str (not "NoneType") to str - traceback: Traceback (most recent call last): File "/config/custom_components/kia_uvo/config_flow.py", line 155, in async_step_user self.token = await self.hass.async_add_executor_job( File "/usr/local/lib/python3.10/concurrent/futures/thread.py", line 58, in run result = self.fn(*self.args, **self.kwargs) File "/config/custom_components/kia_uvo/KiaUvoApiEU.py", line 133, in login ) = self.get_access_token(stamp) File "/config/custom_components/kia_uvo/KiaUvoApiEU.py", line 375, in get_access_token "grant_type=authorization_code&redirect_uri=https%3A%2F%2F" TypeError: can only concatenate str (not "NoneType") to str
kia_uvo Exception in kia_uvo login : 'result' - traceback: Traceback (most recent call last): File "/config/custom_components/kia_uvo/config_flow.py", line 155, in async_step_user self.token = await self.hass.async_add_executor_job( File "/usr/local/lib/python3.10/concurrent/futures/thread.py", line 58, in run result = self.fn(*self.args, **self.kwargs) File "/config/custom_components/kia_uvo/KiaUvoApiCA.py", line 74, in login response = response["result"] KeyError: 'result'

**To Reproduce**
try to log in

**Expected behavior**
Successful log in

**Screenshots**
https://disk.yandex.ru/i/MjWP9Ght8gz8pQ

**Additional context**
I tried to logout and login back in the official app. It doesn't help even after a few days.
I have also tried to create several new accounts, adding my car to these new accounts using the "Share" feature, as well as cleanly adding the car to this new account. It does not help. I am ready to provide everything necessary for testing and solving the problem. Thanks in advance for any help.

**Comment by cdnninja** [comment]
Not by a computer at the moment.  Two things I noticed. 1. You have a canadian and eu error.   Did you add the integration twice? And once select Canada?

Try remove your config and start over. 

Second you say Russian / eu.  I don't know if these are the same region.   I doubt it.  If not this isn't supported.

**Comment by romanovroman74** [comment]
> Not by a computer at the moment. Two things I noticed. 1. You have a canadian and eu error. Did you add the integration twice? And once select Canada?
> 
> Try remove your config and start over.
> 
> Second you say Russian / eu. I don't know if these are the same region. I doubt it. If not this isn't supported.

Yes, in the previous test I tried to log in using different regions.
Yes, indeed, it is possible that Russia and Europe use different servers for the Bluelink service. Is it possible to change the server address somewhere in the code? How can you add support for other regions?

Using WireShark I found out the server my phone is accessing, the address is: 109.167.192.116

If I can't change the server in your integration code, maybe I can redirect requests on my router to the Russian server I need?

I really need this integration...

**Comment by cdnninja** [comment]
Typical the whole API is slightly different.  Just enough to make it so we have to re-write it.   

That IP looks like a different endpoint.   If a website exist for Hyundai this method works for getting API details: https://github.com/Hyundai-Kia-Connect/kia_uvo/wiki/How-to-sniff-API-via-MS-Edge

If not it needs to be android based: https://github.com/Hyundai-Kia-Connect/kia_uvo/wiki/How-to-reverse-engineer-app-traffic-on-Android-phone

I am also going to switch this to a discussion until we have something concrete to work with.
